src/pj_obs_api.c

```diff
@@ -226,60 +285,95 @@ enum pj_log_level pj_log_level (PJ *P, enum pj_log_level log_level) {
 }
 
 
+
+
+
+/*  logging  */
+
+/* pj_vlog resides in pj_log.c and relates to pj_log as vsprintf relates to sprintf */
+void pj_vlog( projCtx ctx, int level, const char *fmt, va_list args );
+
+void proj_log_error (PJ *P, const char *fmt, ...) {
+    va_list args;
+    va_start( args, fmt );
+    pj_vlog (pj_get_ctx (P), PJ_LOG_ERROR , fmt, args);
+    va_end( args );
+}
+
+void proj_log_debug (PJ *P, const char *fmt, ...) {
+    va_list args;
+    va_start( args, fmt );
+    pj_vlog (pj_get_ctx (P), PJ_LOG_DEBUG_MAJOR , fmt, args);
+    va_end( args );
+}
+
+void proj_log_trace (PJ *P, const char *fmt, ...) {
+    va_list args;
+    va_start( args, fmt );
+    pj_vlog (pj_get_ctx (P), PJ_LOG_DEBUG_MINOR , fmt, args);
+    va_end( args );
+}
+
 /* Put a new logging function into P's context. The opaque object app_data is passed as first arg at each call to the logger */
-void pj_log_func (PJ *P, void *app_data, void (*log)(void *, int, const char *)) {
+void proj_log_func (PJ *P, void *app_data, void (*log)(void *, int, const char *)) {
     PJ_CONTEXT *ctx = pj_get_ctx (P);
     ctx->app_data = app_data;
     if (0!=log)
         ctx->logger = log;
 }
 
 
-/* Move P to a new context - initially a copy of the default context */
-int pj_context_renew (PJ *P) {
-    PJ_CONTEXT *ctx = pj_ctx_alloc ();
-    if (0==ctx) {
-        pj_err_level (P, ENOMEM);
-        return 1;
-    }
-
+/* Move P to a new context - or to the default context i 0 is specified */
+void proj_context_set (PJ *P, PJ_CONTEXT *ctx) {
+    if (0==ctx)
+        ctx = pj_get_default_ctx ();
     pj_set_ctx (P, ctx);
-    return 0;
-}
-
-
-/* Move daughter to mother's context */
-void pj_context_inherit (PJ *mother, PJ *daughter) {
-    pj_set_ctx (daughter, pj_get_ctx (mother));
+    return;
 }
 
 
-void pj_context_free (const PJ *P) {
+/* Create a new context - or provide a pointer to the default context */
+/* Prepared for more elaborate sscanf style setup syntax */
+PJ_CONTEXT *proj_context_create (char *setup, ...) {
     PJ_CONTEXT *ctx;
+    va_list args;
+
+    /* The classic single threaded, backwards compatible way */
+    if (0==setup)
+        return pj_get_default_ctx ();
+
+    /* Simplest multi threaded way: ctx is just a container for pj_errno */
+    ctx = pj_ctx_alloc ();
+    if (0==ctx)
+        return 0;
+    if (0==strcmp(setup, ""))
+        return ctx;
+
+    /* Multi threaded with thread specific logging */
+    va_start (args, setup);
```

Comment:
va_list are notoriously hard to use correctly. Just ask users and developers of libtiff TIFFGetField() / TIFFSetField()...  Why not having setters on the PJ_CONTEXT* ?